sync2/atxs.go

```diff
@@ -0,0 +1,326 @@
+			elapsed := h.clock.Since(startTime)
+			h.logger.Debug("fetched atxs",
+				zap.Int("total", total),
+				zap.Int("downloaded", numDownloaded),
+				zap.Float64("rate per sec", float64(numDownloaded)/elapsed.Seconds()))
+		}
+	}
+	return nil
+}
+
+type MultiEpochATXSyncer struct {
+	logger            *zap.Logger
+	oldCfg            Config
+	newCfg            Config
+	parallelLoadLimit int
+	hss               HashSyncSource
+	newEpoch          types.EpochID
+	atxSyncers        []HashSync
+}
+
+func NewMultiEpochATXSyncer(
+	logger *zap.Logger,
+	hss HashSyncSource,
+	oldCfg, newCfg Config,
+	parallelLoadLimit int,
+) *MultiEpochATXSyncer {
+	return &MultiEpochATXSyncer{
+		logger:            logger,
+		oldCfg:            oldCfg,
+		newCfg:            newCfg,
+		parallelLoadLimit: parallelLoadLimit,
+		hss:               hss,
+	}
+}
+
+func (s *MultiEpochATXSyncer) load(newEpoch types.EpochID) error {
+	if len(s.atxSyncers) < int(newEpoch) {
+		s.atxSyncers = append(s.atxSyncers, make([]HashSync, int(newEpoch)-len(s.atxSyncers))...)
+	}
+	s.newEpoch = newEpoch
+	var eg errgroup.Group
+	if s.parallelLoadLimit > 0 {
+		eg.SetLimit(s.parallelLoadLimit)
+	}
+	for epoch := types.EpochID(1); epoch <= newEpoch; epoch++ {
+		if s.atxSyncers[epoch-1] != nil {
+			continue
+		}
+		eg.Go(func() error {
+			name := fmt.Sprintf("atx-sync-%d", epoch)
+			cfg := s.oldCfg
+			if epoch == newEpoch {
+				cfg = s.newCfg
+			}
+			hs := s.hss.CreateHashSync(name, cfg, epoch)
+			if err := hs.Load(); err != nil {
+				return fmt.Errorf("load ATX syncer for epoch %d: %w", epoch, err)
+			}
+			s.atxSyncers[epoch-1] = hs
+			return nil
+		})
+	}
+	return eg.Wait()
+}
+
+// EnsureSync ensures that ATX sync is active for all the epochs up to and including
+// currentEpoch, and that all ATXs are
+// synced up to and including lastWaitEpoch.
+// If newEpoch argument is non-zero, faster but less memory efficient sync is used for
+// that epoch, based on the newCfg (larger maxDepth).
+// For other epochs, oldCfg is used which corresponds to slower but more memory efficient
+// sync (smaller maxDepth).
+// It returns the last epoch that was synced synchronously.
+func (s *MultiEpochATXSyncer) EnsureSync(
+	ctx context.Context,
+	lastWaitEpoch, newEpoch types.EpochID,
+) (lastSynced types.EpochID, err error) {
+	if newEpoch != s.newEpoch && int(s.newEpoch) <= len(s.atxSyncers) && s.newEpoch > 0 {
+		s.atxSyncers[s.newEpoch-1].Stop()
+		s.atxSyncers[s.newEpoch-1] = nil
+	}
+	if err := s.load(newEpoch); err != nil {
+		return lastSynced, err
+	}
+	for epoch := types.EpochID(1); epoch <= newEpoch; epoch++ {
+		syncer := s.atxSyncers[epoch-1]
+		if epoch <= lastWaitEpoch {
+			s.logger.Info("waiting for epoch to sync", zap.Uint32("epoch", epoch.Uint32()))
+			if err := syncer.StartAndSync(ctx); err != nil {
+				return lastSynced, fmt.Errorf("error syncing old ATXs: %w", err)
+			}
+			lastSynced = epoch
+		} else {
+			syncer.Start()
+		}
```

Comment:
NIT: without `else` its clearer to me what's going on:

```suggestion
		if epoch > lastWaitEpoch {
			syncer.Start()
			continue
		}
		s.logger.Info("waiting for epoch to sync", zap.Uint32("epoch", epoch.Uint32()))
		if err := syncer.StartAndSync(ctx); err != nil {
			return lastSynced, fmt.Errorf("error syncing old ATXs: %w", err)
		}
		lastSynced = epoch
```

Reply:
Fixed